### ISSUE TITLE
cdist sometimes returns wrong results when using matrix multiplication

### ISSUE THREAD

**Issue by louis-aptiv** [pytorch/pytorch#42479]
## 🐛 Bug

When using `torch.cdist` in matrix multiplication mode (either by using the flag `compute_mode='use_mm_for_euclid_dist'`, or by using the flag `compute_mode='use_mm_for_euclid_dist_if_necessary'` with enough inputs) results are sometimes completely wrong depending on the input values.

## To Reproduce

Just run this code in Pytorch >= 1.4.0
```
import torch
print(torch.__version__)

# Normal behavior
a = torch.tensor([[5000]]).float()
b = torch.tensor([[6500]]).float()
d1 = torch.cdist(a, b, p=2, compute_mode="use_mm_for_euclid_dist")
d2 = torch.cdist(a, b, p=2, compute_mode="donot_use_mm_for_euclid_dist")

print(a)
print(b)
print(d1)
print(d2)

print()

# Buggy behavior
a = torch.tensor([[512695]]).float()
b = torch.tensor([[512804]]).float()
d1 = torch.cdist(a, b, p=2, compute_mode="use_mm_for_euclid_dist")
d2 = torch.cdist(a, b, p=2, compute_mode="donot_use_mm_for_euclid_dist")

print(a)
print(b)
print(d1)
print(d2)
```

## Expected behavior

Current output I get is the following:
```
1.6.0
tensor([[5000.]])
tensor([[6500.]])
tensor([[1500.]])
tensor([[1500.]])

tensor([[512695.]])
tensor([[512804.]])
tensor([[0.]])
tensor([[109.]])
```

Expected output is the following:
```
1.6.0
tensor([[5000.]])
tensor([[6500.]])
tensor([[1500.]])
tensor([[1500.]])

tensor([[512695.]])
tensor([[512804.]])
tensor([[109.]])
tensor([[109.]])
```

## Environment

PyTorch version: 1.6.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14.6
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.14.5
[pip3] pytorch-ignite==0.2.0
[pip3] torch==1.6.0
[pip3] torchvision==0.7.0
[conda] numpy                     1.14.5                   pypi_0    pypi
[conda] pytorch-ignite            0.2.0                    pypi_0    pypi
[conda] torch                     1.6.0                    pypi_0    pypi
[conda] torchvision               0.7.0                    pypi_0    pypi

**Comment by agolynski** [comment]
@vadimkantorov There must be some kind of loss of precision here.

```
a = torch.tensor([[512695]]).double()
b = torch.tensor([[512804]]).double()
d1 = torch.cdist(a, b, p=2, compute_mode="use_mm_for_euclid_dist")
d2 = torch.cdist(a, b, p=2, compute_mode="donot_use_mm_for_euclid_dist")

```

and works fine.

**Comment by louis-aptiv** [comment]
Thank you @agolynski, I tested the `.double` version and it seems to also have some problems (which are probably the same problems as in #37734), although not as critical:
```
a = torch.tensor([[386784556587]]).double()
b = torch.tensor([[386783820152]]).double()
d1 = torch.cdist(a, b, p=2, compute_mode="use_mm_for_euclid_dist")
d2 = torch.cdist(a, b, p=2, compute_mode="donot_use_mm_for_euclid_dist")

print(a)
print(b)
print(a - b)
print(d1)
print(d2)
```
which gave me this output:
```
tensor([[3.8678e+11]], dtype=torch.float64)
tensor([[3.8678e+11]], dtype=torch.float64)
tensor([[736435.]], dtype=torch.float64)
tensor([[736448.9539]], dtype=torch.float64)
tensor([[736435.]], dtype=torch.float64)
```

**Comment by ngimel** [comment]
This is an expected behavior, given that mm distance is calculated as a^2-2*a*b+b^2. If a and b are close, then catastrophic cancellation happens and result is imprecise (can even be negative, but we clamp it to be non-negative). 
Using matrix multiplication dramatically improves performance, but if accuracy is not enough for your application, then you should not use it.

**Comment by playertr** [comment]
@ngimel The tradeoff between `cdist` precision and speed was not apparent to me from the [documentation](https://pytorch.org/docs/stable/generated/torch.cdist.html) for this function, which does not explain why one might prefer to use mm distance and its implications for numerical stability.

I don't yet fully understand the source of the numerical imprecision, but it appears that folks have [discussed](https://discuss.pytorch.org/t/efficient-distance-matrix-computation/9065) the tradeoffs for batched distance computation in the Pytorch forums before.

Would improving the `cdist` documentation to better explain the `compute_mode` tradeoffs be a welcome contribution for a first pull request?

**Comment by ngimel** [comment]
Yes, improving documentation is always welcome!

**Comment by BensonRen** [comment]
> @ngimel The tradeoff between `cdist` precision and speed was not apparent to me from the [documentation](https://pytorch.org/docs/stable/generated/torch.cdist.html) for this function, which does not explain why one might prefer to use mm distance and its implications for numerical stability.
> 
> I don't yet fully understand the source of the numerical imprecision, but it appears that folks have [discussed](https://discuss.pytorch.org/t/efficient-distance-matrix-computation/9065) the tradeoffs for batched distance computation in the Pytorch forums before.
> 
> Would improving the `cdist` documentation to better explain the `compute_mode` tradeoffs be a welcome contribution for a first pull request?

It is still not clear enough frankly... I strongly recommend adding extra information in the documentation for that
